### PR TITLE
probe:  ProbePointsHelper enhancements

### PR DESCRIPTION
**Pull request by Arksine**
- Extract Lift-Z code into its own function, and lift z prior to moving to the first point.
- Add point multi-sampling to ProbePointsHelper.  Initially I was going to add this directly to bed_mesh by duplicating points during generation and averaging the returned values.  There are two advantages of adding this to ProbePointsHelper: it is functional for all calibration classes and it allows the user to specify a different retraction distance between samples.
- Basic support for a menu based wizard during manual calibration.  I noticed you mention this in the IRC channel and I thought I would take a shot at it.  I had to extend menu.py to get this to work, and while  I'm confident I didn't break anything, @mcmatrix may want to review my changes to be sure.  

Currently the wizard allows the user to move Z in .1 steps with realtime mode enabled.  I looked at ways to allow the user to select step resolution (1mm, .1mm, and .01mm), but the only way I could come up with was to add menu items for each resolution.   I felt like this looked over-burdensome on the wizard, so I removed 1mm and .01mm, however it wouldn't be any problem to add them back if you think it is necessary as I left them in menu.cfg.

Signed-off-by:  Eric Callahan <arksine.code@gmail.com>

### CODE REVIEW

**Comment by mcmatrix** [comment]
@Arksine How this `realtime` mode is working for you.
One user reported that with longer moves (steps) it
starts acting weird. After some distance, the axis starts to oscillate randomly up and down.
When using small steps it's more or less ok.

This is oscillating when doing long moves
```
[menu __control __move_10mm __axis_x]
type: input
name: "X:{0:05.1f} "
parameter: toolhead.xpos
input_min: -41
input_max: 250.0
input_step: 10.0
realtime: true
gcode: G1 X{0:.1f}

```

**Comment by Arksine** [comment]
@mcmatrix I haven't had any issue with it in my testing, but I have only been using .1mm steps.  I'll see how it does with 1mm steps.  This sounds like its possible that gcodes are getting executed out of order?

File: klippy/extras/probe.py [hunk 1]
```diff
@@ -180,15 +183,31 @@ def get_last_xy_home_positon(self):
...
         except homing.EndstopError as e:
             self.finalize(False)
             raise self.gcode.error(str(e))
+    def probe_point(self):
+        avg_z = 0.
+        for i in range(self.samples):
+            self.gcode.run_script_from_command("PROBE")
+            self.toolhead.wait_moves()
+            kin_pos = self.callback.get_probed_position()
+            avg_z += kin_pos[2]
```
Comment (KevinOConnor) [inline klippy/extras/probe.py]: I believe this would break delta calibration.  The delta calibration returns a "stable position" (which is the number of steps taken since hitting the endstop on each tower).  At a high-level I was thinking that get_probed_position() would return a "cookie" that the ProbePointsHelper class would just pass to finalize().

Could the additional samples just be passed to finalize() and let finalize() do the averaging?  (The other users of ProbePointsHelper will do that automatically.)

File: klippy/extras/probe.py [hunk 2]
```diff
@@ -224,9 +243,10 @@ def move_next(self):
         self.gcode.reset_last_position()
     cmd_NEXT_help = "Move to the next XY position to probe"
     def cmd_NEXT(self, params):
-        # Record current position
-        self.toolhead.wait_moves()
-        self.results.append(self.callback.get_probed_position())
+        if not self.probe:
```
Comment (KevinOConnor) [inline klippy/extras/probe.py]: Is this change necessary?  (It's not a problem to check for self.probe, but wouldn't NEXT only ever be called when self.probe is None.)

Separately, as a random side-note, it may be safer in python to explicitly check for `if some_class is None:` then to check for truth (`if some_class:`), as some classes may end up implementing the truth callback.

Comment (Arksine) [reply klippy/extras/probe.py]: Hi Kevin, unless I am missing something, self.cmd_NEXT({}) is called on line 249 in the busy loop.  If I don't make the check then an extra point will be appended after the probe_point() function completes.

I did make the suggested change to the truth check.

**Comment by Arksine** [comment]
I have pushed an updated version that does the averaging in the finalize() callback.  If the self.cmd_NEXT() check is whooshing on me I'll take it out, but I'm reasonably certain not having it will cause an issue.

**Comment by KevinOConnor** [comment]
Thanks.  I committed the first two patches.  I'm interested in getting @mcmatrix's feedback on the third patch.  (I also haven't had a chance to look at it closely.)  Thanks for working on it.

-Kevin

**Comment by Arksine** [comment]
Great, thanks!  I'm definitely interested in his feedback as well.  

Also, I have a theory on why the menu's realtime mode could cause oscillation.  The menu intializes the input value using the toolhead class.  If the user has a gcode transform active this value won't reflect the actual toolhead position while editing, as the input value isn't reset until the user has finished editing.  When I get a chance this evening I'll test this theory to see what happens.  @mcmatrix Is there an issue for this I can comment on?

**Comment by mcmatrix** [comment]
Thanks, guys.
I've been busy this week. I'll try to review the menu related part tomorrow evening.
@Arksine I'll create an issue for this real-time problem.
I am looking forward to the results of your test.

**Comment by mcmatrix** [comment]
@Arksine Thanks! Nice work. Clever way to get UI for functionality.
Unfortunately, I realized that it was impossible for me to test it completely.
I don't have fully working Klipper printer (My Anycubic Kossel is still running Marlin1.1.9). 
For menu testing, I'm using old raspberry and just Ramps with a display. 
It's enough for basic menu test but not probe and movements related functionality.
Default menu and decks looked okay. 
I cannot reach to your special menus without Klipper on the real printer.
Please continue with your great work. Sorry for the delay.

**Comment by mcmatrix** [comment]
I now had a chance to use `DELTA_CALIBRATION` with your menu.
Looks very nice.
Some of my observations.

- I personally would like to have also 1mm and .01mm menus
Somehow I felt when making close adjustments with .1mm that it was too coarse.
Additional items will fit nicely into 4 rows but you have to consider 16cols screen too.
```
>Move 1.0 Z25.00
>Move 0.1 Z25.00
>Move .01 Z25.00
>Move Next
```

- When I have `menu_timeout` set then it will go back from Move submenu to `Cal Wizard` card.
That status line `Moving...` is static or should it reflect what is going on at the moment.
By using `long press` I'm able to go back `Move` submenu.
```
Cal Wizard:
Moving...
X:0.00 Y:0.00
```


- Can I disable / enable this menu in config for probing? It could be that even I have a display I prefer to use a terminal commands or maybe my own menu.

- Should it leave ProbePointsHelper into parameters after running or should it clean up?
I dont know, it probably wont harm anyone to leave it there.

Let you know if I'll find more

**Comment by mcmatrix** [comment]
At the moment I'm testing new dev for encoder fast rate stepping.
If it will work then there's no need have that many `Move` menu items.
For example, then normal speed step size can be `0.01` and fast rate `0.1`

**Comment by Arksine** [comment]
Thanks for looking at this @mcmatrix.  A few responses to your observations:

> When I have menu_timeout set then it will go back from Move submenu to Cal Wizard card.
> That status line Moving... is static or should it reflect what is going on at the moment.
> By using long press I'm able to go back Move submenu.

The Moving screen is designed to be static.  Its simply a status screen shown when moving to the next point.  I hadn't considered what a timeout would do, my assumption is that the user wouldn't typically leave the move submenu inactive long enough for the timeout to trigger.

> Can I disable / enable this menu in config for probing? It could be that even I have a display I prefer to use a terminal commands or maybe my own menu.

We could probably add an option to disable this.  FWIW, I tested this functionality with terminal gcodes and they work as well.  So while the display would change, it doesn't remove the ability to use octoprint instead of the encoder.

> Should it leave ProbePointsHelper into parameters after running or should it clean up?
> I dont know, it probably wont harm anyone to leave it there.

The finalize() function does reset the write_enabled parameter, however it doesn't do anything with next_pos.  I don't think it should be necessary to do anything with it.

> I also noticed that the automatic probe behaved a little bit differently when in master.
> Speed was different than in master.

Yes, a more recent commit in master fixed the speed to better work with Delta printers.

> At the moment I'm testing new dev for encoder fast rate stepping.
> If it will work then there's no need have that many Move menu items.
> For example, then normal speed step size can be 0.01 and fast rate 0.1

Sounds great!  This would be a very useful addition.

A final thought:  I hardcoded the menu functionality into ProbePointsHelper for two reasons:
* I wanted to toggle the screen between status and input in real time based on the ProbePointsHelper state.
* I had thought that I should automatically launch the wizard if the menu is available.   

If we want to make it optional, we could instead make the wizard launch when a calibration item is selected via the menu.  However, to get the menu to respond to state in real time I believe some changes will need to be made to the way menu lists behave.  Based on my testing, it seems that only menu cards respond to the 'enable' option in real time.  If we can update lists in the same manner then it would be possible to switch between status and move screens based on the enabled state.

**Comment by mcmatrix** [comment]
> Based on my testing, it seems that only menu cards respond to the 'enable' option in real time. 

That's true, I had to remove such "realtime" enable from the normal menu. It's was causing problems for the menu navigation when items can hide and appear in real-time.

**Comment by mcmatrix** [comment]
I was looking at your menu implementation and I think it can be simplified 
to 2 menu cards without the need for longpress_menu and list item.
At any time only one card is active.
When moving then `card2` is active and when user should adjust Z then `card1`.
Then user will be all the time on right screen.

To achieve it, I need to add change in menu code.
Currently, only selection indicator for card item and menu-group item is blinking.
I'll add the possibility to use a cursor.

It's just idea! Maybe it will not work but let's see how it goes. 
I'll report back when I have results.

I think for `Move Z` normal step size can be with high precision `0.01` and
fast rate step2 more coarse `0.1`

```
card1
card2
1234567890123456
----------------
= Probe helper =
Please adjust Z
>Move Z:000.000
>Probe next
----------------
= Probe helper =
Moving to point
X:100.0 Y:100.0
Please wait...
----------------
```
What do you think about this?

**Comment by Arksine** [comment]
@mcmatrix This sounds good!  Considering that the remaining items in this PR are somewhat stale, I was going to close the request and think about other approaches.  Your approach sounds like it will work.  I'll leave this open until you report on your findings.  If things work well we can use your implementation I'll close this.

**Comment by Arksine** [comment]
Closing this PR as new, improved version of this work is now in PR #741
